[FEATURE] Streamlines Renderer for vector dataset on mesh layer.

vcloarec opened the pull request:
fixes #29534 

The vector dataset of a mesh layer is currently only rendered with arrows. 

This PR adds a new feature to render vectors field with streamlines.

The streamlines are seeded with a start point. The seeding can be done from the vertices of the mesh, from a user grid or randomly. The user chooses a method in a combo box: `On Mesh/grid` or  `Randomly`.
With the option `On Mesh/Grid`, by default, the vertices of the mesh are used to start the streamlines, but the user can define a grid (as for arrows) by entering spacing (px).
An alternative is to choose `Randomly`. The start points are defined randomly and 
the user can then define a density that represents the part of pixels occupied by streamlines in the extent of the mesh displayed on the screen.

![streamlines](https://user-images.githubusercontent.com/7416892/69353527-055dac80-0c55-11ea-9afd-2d0a38c3c8da.gif)


- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [x] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

PeterPetrik commented:
fixes #29534

PeterPetrik commented on src/core/mesh/qgstriangularmesh.h:
```diff
@@ -137,6 +148,30 @@ namespace QgsMeshUtils
+   *          = 0 if on the line
+   *          < 0 if on the right
+   *  NB : triangle surface = Isleft / 2
+   *  * \since QGIS 3.12
```
extra *

PeterPetrik commented on src/core/mesh/qgstriangularmesh.h:
```diff
@@ -137,6 +148,30 @@ namespace QgsMeshUtils
+
+  /**
+   * Test if point p is on the 2D triangle
+   *  * \since QGIS 3.12
```
extra *

PeterPetrik commented on src/core/mesh/qgstriangularmesh.h:
```diff
@@ -137,6 +148,30 @@ namespace QgsMeshUtils
+
+  /**
+   * Test if point p is on the face defined with vertices
+   *  * \since QGIS 3.12
```
extra *

PeterPetrik commented on src/core/mesh/qgsmeshvectorrenderer.h:
```diff
@@ -42,26 +71,26 @@ class QgsRenderContext;
  * \since QGIS 3.2
  */
-class QgsMeshVectorRenderer
+class CORE_EXPORT QgsMeshVectorArrowRenderer : public QgsMeshVectorRenderer
```
does it need to be exported outside qgis_core?

PeterPetrik commented on src/core/mesh/qgsmeshvectorrenderer.h:
```diff
@@ -34,6 +34,35 @@ class QgsRenderContext;
 ///@cond PRIVATE
 
+
+class CORE_EXPORT QgsMeshVectorRenderer
```
does it need to be exported outside qgis_core?

PeterPetrik commented on src/core/mesh/qgsmeshtracerenderer.h:
```diff
@@ -0,0 +1,279 @@
+ * \note not available in Python bindings
+ * \since QGIS 3.12
+ */
+class CORE_EXPORT QgsMeshStreamField
```
does it need to be exported outside qgis_core?

PeterPetrik reviewed (commented):
good work

vcloarec committed:
[FEATURE] Streamlines fixes #29534

The vector field data set in a mesh layer is currently only renderered with arrows. This PR add a new feature to render vector field with streamlines.
The streamlines are seeded with a start point. The seeding can be done from the vertices of the mesh, from a used grid or randomly.

vcloarec committed:
typo
